frontend/ui/wikipedia.lua

```diff
@@ -95,84 +95,36 @@ function Wikipedia:getWikiServer(lang)
-    require('socket.http').USERAGENT:gsub(" ", "/") )
-
--- Codes that getUrlContent may get from requester.request()
-local TIMEOUT_CODE = "timeout" -- from socket.lua
-local MAXTIME_CODE = "maxtime reached" -- from sink_table_with_maxtime
-
--- Sink that stores into a table, aborting if maxtime has elapsed
-local function sink_table_with_maxtime(t, maxtime)
-    -- Start counting as soon as this sink is created
-    local start_secs, start_usecs = ffiutil.gettime()
-    local starttime = start_secs + start_usecs/1000000
-    t = t or {}
-    local f = function(chunk, err)
-        local secs, usecs = ffiutil.gettime()
-        if secs + usecs/1000000 - starttime > maxtime then
-            return nil, MAXTIME_CODE
-        end
-        if chunk then table.insert(t, chunk) end
-        return 1
-    end
-    return f, t
-end
-
 -- Get URL content
 local function getUrlContent(url, timeout, maxtime)
-    local socket = require('socket')
-    local ltn12 = require('ltn12')
-    local http = require('socket.http')
-    local https = require('ssl.https')
-
-    local requester
-    if url:sub(1,7) == "http://" then
-        requester = http
-    elseif url:sub(1,8) == "https://" then
-        requester = https
-    else
+    local http = require("socket.http")
+    local ltn12 = require("ltn12")
+    local socket = require("socket")
+    local socketutil = require("socketutil")
+    local socket_url = require("socket.url")
+
+    local parsed = socket_url.parse(url)
+    if parsed.scheme ~= "http" and parsed.scheme ~= "https" then
         return false, "Unsupported protocol"
     end
     if not timeout then timeout = 10 end
-    -- timeout needs to be set to 'http', even if we use 'https'
-    http.TIMEOUT, https.TIMEOUT = timeout, timeout
 
-    local request = {}
     local sink = {}
-    request['url'] = url
-    request['method'] = 'GET'
-    request['headers'] = {
-        ["User-Agent"] = USER_AGENT,
+    socketutil:set_timeout(timeout, maxtime or 30)
+    local request = {
+        url     = url,
+        method  = "GET",
+        sink    = maxtime and socketutil.table_sink(sink) or ltn12.sink.table(sink),
     }
-    -- 'timeout' delay works on socket, and is triggered when
-    -- that time has passed trying to connect, or after connection
-    -- when no data has been read for this time.
-    -- On a slow connection, it may not be triggered (as we could read
-    -- 1 byte every 1 second, not triggering any timeout).
-    -- 'maxtime' can be provided to overcome that, and we start counting
-    -- as soon as the first content byte is received (but it is checked
-    -- for only when data is received).
-    -- Setting 'maxtime' and 'timeout' gives more chance to abort the request when
-    -- it takes too much time (in the worst case: in timeout+maxtime seconds).
-    -- But time taken by DNS lookup cannot easily be accounted for, so
-    -- a request may (when dns lookup takes time) exceed timeout and maxtime...
```

Comment:
Nothing worth keeping from my comment ? :)
(To keep a bit of higher level/practical info/context ? There is mostly only low-level C/socket technical details in socketutil :/)

Reply:
That's similar in spirit with the comment I added in the last few commits in set_timeout ;).

Reply:
Expanded on said comments after looking into what the hell happened with the socket timeouts ;).

```diff
diff --git a/src/buffer.c b/src/buffer.c
index ac5c531..d1f5ddd 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -76,6 +76,7 @@ int buffer_meth_send(lua_State *L, p_buffer buf) {
     const char *data = luaL_checklstring(L, 2, &size);
     long start = (long) luaL_optnumber(L, 3, 1);
     long end = (long) luaL_optnumber(L, 4, -1);
+    printf("timeout_markstart in buffer_meth_send\n");
     timeout_markstart(buf->tm);
     if (start < 0) start = (long) (size+start+1);
     if (end < 0) end = (long) (size+end+1);
@@ -107,6 +108,7 @@ int buffer_meth_receive(lua_State *L, p_buffer buf) {
     luaL_Buffer b;
     size_t size;
     const char *part = luaL_optlstring(L, 3, "", &size);
+    printf("timeout_markstart in buffer_meth_receive\n");
     timeout_markstart(buf->tm);
     /* initialize buffer with optional extra prefix
      * (useful for concatenating previous partial results) */
diff --git a/src/inet.c b/src/inet.c
index ec73fea..c3f8d6c 100644
--- a/src/inet.c
+++ b/src/inet.c
@@ -399,6 +399,7 @@ const char *inet_tryconnect(p_socket ps, int *family, const char *address,
         return err;
     }
     for (iterator = resolved; iterator; iterator = iterator->ai_next) {
+        printf("timeout_markstart in inet:inet_tryconnect\n");
         timeout_markstart(tm);
         /* create new socket if necessary. if there was no
          * bind, we need to create one for every new family
diff --git a/src/luasocket.c b/src/luasocket.c
index 0fd99f7..835c633 100755
--- a/src/luasocket.c
+++ b/src/luasocket.c
@@ -96,7 +96,7 @@ static int base_open(lua_State *L) {
 /*-------------------------------------------------------------------------*\
 * Initializes all library modules.
 \*-------------------------------------------------------------------------*/
-LUASOCKET_API int luaopen_socket_core(lua_State *L) {
+LUASOCKET_API int luaopen_socket_score(lua_State *L) {
     int i;
     base_open(L);
     for (i = 0; mod[i].name; i++) mod[i].func(L);
diff --git a/src/luasocket.h b/src/luasocket.h
index d22b1be..c524642 100644
--- a/src/luasocket.h
+++ b/src/luasocket.h
@@ -31,6 +31,6 @@
 /*-------------------------------------------------------------------------*\
 * Initializes the library.
 \*-------------------------------------------------------------------------*/
-LUASOCKET_API int luaopen_socket_core(lua_State *L);
+LUASOCKET_API int luaopen_socket_score(lua_State *L);
 
 #endif /* LUASOCKET_H */
diff --git a/src/makefile b/src/makefile
index 522d378..8e26231 100755
--- a/src/makefile
+++ b/src/makefile
@@ -175,7 +175,7 @@ O_linux=o
 CC_linux=gcc
 DEF_linux=-DLUASOCKET_$(DEBUG)
 CFLAGS_linux=$(LUAINC:%=-I%) $(DEF) -Wall -Wshadow -Wextra \
-       -Wimplicit -O2 -ggdb3 -fpic
+       -Wimplicit -fpic
 LDFLAGS_linux=-O -shared -fpic -o 
 LD_linux=gcc
 SOCKET_linux=usocket.o
@@ -188,7 +188,7 @@ O_freebsd=o
 CC_freebsd=gcc
 DEF_freebsd=-DLUASOCKET_$(DEBUG) -DUNIX_HAS_SUN_LEN
 CFLAGS_freebsd=$(LUAINC:%=-I%) $(DEF) -Wall -Wshadow -Wextra \
-       -Wimplicit -O2 -ggdb3 -fpic
+       -Wimplicit -fpic
 LDFLAGS_freebsd=-O -shared -fpic -o 
 LD_freebsd=gcc
 SOCKET_freebsd=usocket.o
@@ -201,7 +201,7 @@ O_solaris=o
 CC_solaris=gcc
 DEF_solaris=-DLUASOCKET_$(DEBUG)
 CFLAGS_solaris=$(LUAINC:%=-I%) $(DEF) -Wall -Wshadow -Wextra \
-       -Wimplicit -O2 -ggdb3 -fpic
+       -Wimplicit -fpic
 LDFLAGS_solaris=-lnsl -lsocket -lresolv -O -shared -fpic -o 
 LD_solaris=gcc
 SOCKET_solaris=usocket.o
@@ -415,9 +415,9 @@ install:
        $(INSTALL_DIR) $(INSTALL_SOCKET_LDIR)
        $(INSTALL_DATA) $(TO_SOCKET_LDIR) $(INSTALL_SOCKET_LDIR)
        $(INSTALL_DIR) $(INSTALL_SOCKET_CDIR)
-       $(INSTALL_EXEC) $(SOCKET_SO) $(INSTALL_SOCKET_CDIR)/core.$(SO)
+       $(INSTALL_EXEC) $(SOCKET_SO) $(INSTALL_SOCKET_CDIR)/score.$(SO)
        $(INSTALL_DIR) $(INSTALL_MIME_CDIR)
-       $(INSTALL_EXEC) $(MIME_SO) $(INSTALL_MIME_CDIR)/core.$(SO)
+       $(INSTALL_EXEC) $(MIME_SO) $(INSTALL_MIME_CDIR)/mcore.$(SO)
 
 install-unix: install
        $(INSTALL_EXEC) $(UNIX_SO) $(INSTALL_SOCKET_CDIR)/$(UNIX_SO)
diff --git a/src/mime.c b/src/mime.c
index 6c210e4..1350c0c 100755
--- a/src/mime.c
+++ b/src/mime.c
@@ -167,7 +167,7 @@ static const UC b64unbase[] = {
 /*-------------------------------------------------------------------------*\
 * Initializes module
 \*-------------------------------------------------------------------------*/
-LUASOCKET_API int luaopen_mime_core(lua_State *L)
+LUASOCKET_API int luaopen_mime_mcore(lua_State *L)
 {
     lua_newtable(L);
     luaL_setfuncs(L, func, 0);
diff --git a/src/mime.h b/src/mime.h
index 4d938f4..1fc55c3 100644
--- a/src/mime.h
+++ b/src/mime.h
@@ -17,6 +17,6 @@
 #define MIME_COPYRIGHT  "Copyright (C) 2004-2013 Diego Nehab"
 #define MIME_AUTHORS    "Diego Nehab"
 
-LUASOCKET_API int luaopen_mime_core(lua_State *L);
+LUASOCKET_API int luaopen_mime_mcore(lua_State *L);
 
 #endif /* MIME_H */
diff --git a/src/mime.lua b/src/mime.lua
index d3abac5..40f3b94 100644
--- a/src/mime.lua
+++ b/src/mime.lua
@@ -9,7 +9,7 @@
 -----------------------------------------------------------------------------
 local base = _G
 local ltn12 = require("ltn12")
-local mime = require("mime.core")
+local mime = require("mime.mcore")
 local string = require("string")
 local _M = mime
 
diff --git a/src/select.c b/src/select.c
index bb47c45..4969fa1 100644
--- a/src/select.c
+++ b/src/select.c
@@ -65,6 +65,7 @@ static int global_select(lua_State *L) {
     ndirty = check_dirty(L, 1, rtab, &rset);
     t = ndirty > 0? 0.0: t;
     timeout_init(&tm, t, -1);
+    printf("timeout_markstart in select:global_select\n");
     timeout_markstart(&tm);
     ret = socket_select(max_fd+1, &rset, &wset, NULL, &tm);
     if (ret > 0 || ndirty > 0) {
diff --git a/src/socket.lua b/src/socket.lua
index d1c0b16..ea7c8ef 100644
--- a/src/socket.lua
+++ b/src/socket.lua
@@ -9,7 +9,7 @@
 local base = _G
 local string = require("string")
 local math = require("math")
-local socket = require("socket.core")
+local socket = require("socket.score")
 
 local _M = socket
 
diff --git a/src/tcp.c b/src/tcp.c
index 5876bfb..4d268d6 100644
--- a/src/tcp.c
+++ b/src/tcp.c
@@ -212,6 +212,7 @@ static int meth_dirty(lua_State *L)
 static int meth_accept(lua_State *L)
 {
     p_tcp server = (p_tcp) auxiliar_checkclass(L, "tcp{server}", 1);
+    printf("timeout_markstart in tcp:meth_accept\n");
     p_timeout tm = timeout_markstart(&server->tm);
     t_socket sock;
     const char *err = inet_tryaccept(&server->sock, server->family, &sock, tm);
@@ -272,6 +273,7 @@ static int meth_connect(lua_State *L) {
     connecthints.ai_socktype = SOCK_STREAM;
     /* make sure we try to connect only to the same family */
     connecthints.ai_family = tcp->family;
+    printf("timeout_markstart in tcp:meth_connect\n");
     timeout_markstart(&tcp->tm);
     err = inet_tryconnect(&tcp->sock, &tcp->family, address, port,
         &tcp->tm, &connecthints);
diff --git a/src/timeout.c b/src/timeout.c
index 2bdc069..c2c752c 100644
--- a/src/timeout.c
+++ b/src/timeout.c
@@ -95,13 +95,20 @@ double timeout_getretry(p_timeout tm) {
         return -1;
     } else if (tm->block < 0.0) {
         double t = tm->total - timeout_gettime() + tm->start;
-        return MAX(t, 0.0);
+        double r = MAX(t, 0.0);
+        printf("timeout_getretry b < 0: %.6f\n", r);
+        return r;
     } else if (tm->total < 0.0) {
         double t = tm->block - timeout_gettime() + tm->start;
-        return MAX(t, 0.0);
+        double r = MAX(t, 0.0);
+        printf("timeout_getretry t < 0: %.6f\n", r);
+        return r;
     } else {
         double t = tm->total - timeout_gettime() + tm->start;
-        return MIN(tm->block, MAX(t, 0.0));
+        printf("timeout_getretry t: %.6f\n", t);
+        double r = MIN(tm->block, MAX(t, 0.0));
+        printf("timeout_getretry: %.6f\n", r);
+        return r;
     }
 }
 
@@ -112,6 +119,7 @@ double timeout_getretry(p_timeout tm) {
 \*-------------------------------------------------------------------------*/
 p_timeout timeout_markstart(p_timeout tm) {
     tm->start = timeout_gettime();
+    printf("timeout_markstart: %.6f\n", tm->start);
     return tm;
 }
 
@@ -135,7 +143,9 @@ double timeout_gettime(void) {
     struct timeval v;
     gettimeofday(&v, (struct timezone *) NULL);
     /* Unix Epoch time (time since January 1, 1970 (UTC)) */
-    return v.tv_sec + v.tv_usec/1.0e6;
+    double t = v.tv_sec + v.tv_usec/1.0e6;
+    printf("timeout_gettime: %.6f\n", t);
+    return t;
 }
 #endif
 
diff --git a/src/usocket.c b/src/usocket.c
index acfe186..be7b545 100644
--- a/src/usocket.c
+++ b/src/usocket.c
@@ -32,6 +32,7 @@ int socket_waitfd(p_socket ps, int sw, p_timeout tm) {
     if (timeout_iszero(tm)) return IO_TIMEOUT;  /* optimize timeout == 0 case */
     do {
         int t = (int)(timeout_getretry(tm)*1e3);
+        printf("timeout_getretry in socket_waitfd: %d\n", t);
         ret = poll(&pfd, 1, t >= 0? t: -1);
     } while (ret == -1 && errno == EINTR);
     if (ret == -1) return errno;
```

Reply:
In practice, this looks like a lot of this:

```
19:13:09.164778 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.001242>
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.166558
timeout_markstart: 1615572789.166558
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.167348
timeout_markstart: 1615572789.167348
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.167636
timeout_markstart: 1615572789.167636
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.168132
timeout_markstart: 1615572789.168132
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.168954
timeout_markstart: 1615572789.168954
timeout_gettime: 1615572789.169150
timeout_getretry t: 9.999804
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.169379 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.000334>
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.170411
timeout_markstart: 1615572789.170411
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.170743
timeout_markstart: 1615572789.170743
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.171077
timeout_markstart: 1615572789.171077
timeout_gettime: 1615572789.171269
timeout_getretry t: 9.999808
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.171496 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.000404>
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.172409
timeout_markstart: 1615572789.172409
timeout_gettime: 1615572789.172616
timeout_getretry t: 9.999793
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.172841 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.000449>
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.173992
timeout_markstart: 1615572789.173992
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.176001
timeout_markstart: 1615572789.176001
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.176391
timeout_markstart: 1615572789.176391
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.176683
timeout_markstart: 1615572789.176683
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.177011
timeout_markstart: 1615572789.177011
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.177340
timeout_markstart: 1615572789.177340
timeout_gettime: 1615572789.177527
timeout_getretry t: 9.999813
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.177756 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.001846>
timeout_gettime: 1615572789.179977
timeout_getretry t: 9.997363
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.180224 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.000671>
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.181437
timeout_markstart: 1615572789.181437
timeout_gettime: 1615572789.181646
timeout_getretry t: 9.999791
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.181877 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.002694>
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.185108
timeout_markstart: 1615572789.185108
timeout_gettime: 1615572789.185688
timeout_getretry t: 9.999420
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.186267 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.000042>
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.187047
timeout_markstart: 1615572789.187047
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.187403
timeout_markstart: 1615572789.187403
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.187772
timeout_markstart: 1615572789.187772
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.188104
timeout_markstart: 1615572789.188104
timeout_markstart in buffer_meth_receive
timeout_gettime: 1615572789.188517
timeout_markstart: 1615572789.188517
timeout_gettime: 1615572789.188712
timeout_getretry t: 9.999805
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.188939 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.000777>
timeout_gettime: 1615572789.190230
timeout_getretry t: 9.998287
timeout_getretry: 5.000000
timeout_getretry in socket_waitfd: 5000
19:13:09.190481 [2ac7831c] poll([{fd=4, events=POLLIN}], 1, 5000) = 1 ([{fd=4, revents=POLLIN}]) <0.000549>
```

(So, hell, the timeout can be reset *multiple* times between polls).